### PR TITLE
Allow access to child providers in ChainedConfigurationProvider

### LINKED ISSUE #44486: Allow access to child providers in ChainedConfigurationProvider
### API Proposal

This API proposal provides the ability to enumerate a `ChainedConfigurationProvider`'s child providers:

```c#
namespace Microsoft.Extensions.Configuration
{
    public class ChainedConfigurationProvider : IConfigurationProvider, IDisposable
    {
+        /// <summary>
+        /// The chained <see cref="IConfigurationProvider"/> instances.
+        /// </summary>
+        IEnumerable<IConfigurationProvider>? Providers { get; }
```

### Motivation

Developers tend to add their configuration source via a ChainedConfigurationSource/Provider and they would want to access it from the outer `ConfigurationRoot`. 

`ConfigurationRoot` exposed `Providers`. With this new API now, configuration providers that are nested under a `ChainedConfigurationProvider` if the user calls `IConfigurationBulder.AddConfiguration(...)` would now also be able to be accessed. 

### Usage

Sample usage can be found in [AzureAppConfigurationRefresherProvider](https://github.com/Azure/AppConfiguration-DotnetProvider/blob/96abfe3af854bde6d92860e74ab59866c1df03ca/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationRefresherProvider.cs#L21-L27) under the [Azure/AppConfiguration-DotnetProvider](https://github.com/Azure/AppConfiguration-DotnetProvider/) repo.

### Original Description

We have an [issue](https://github.com/Azure/AppConfiguration-DotnetProvider/issues/168) in Azure App Configuration where `AzureAppConfigurationProvider` could be nested under a `ChainedConfigurationProvider` if the user calls `IConfigurationBuilder.AddConfiguration(...)` to add configuration built from the App Configuration provider.

Sample code snippet of this use case:

```cs
public static IWebHost BuildWebHost(string[] args)
{
    var configBuilder = new ConfigurationBuilder();
    IConfiguration configuration = configBuilder.AddAzureAppConfiguration(options =>
    {
        options.Connect(connectionString)
        .ConfigureRefresh(refreshOptions =>
        {
            refreshOptions.Register("Settings:BackgroundColor", refreshAll: true)
                          .SetCacheExpiration(TimeSpan.FromSeconds(10));
        });
    })
    .Build();

    return WebHost.CreateDefaultBuilder(args)
                  .ConfigureAppConfiguration((hostingContext, config) =>
                  {
                      config.AddJsonFile("appsettings.json").Build();
                      config.AddConfiguration(configuration);
                  })
                  .UseStartup<Startup>()
                  .Build();
}
```

Unfortunately, ChainedConfigurationProvider does not provide a way for others to access its child providers. This caused the issue that the App Configuration middleware could not find the instance of AzureAppConfigurationProvider (because in this case its under ChainedConfigurationProvider) and thus cannot obtain the instance of IConfigurationRefresher, which is responsible for refreshing the application configuration ([our relevant middleware code here](https://github.com/Azure/AppConfiguration-DotnetProvider/blob/96abfe3af854bde6d92860e74ab59866c1df03ca/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationRefresherProvider.cs#L21-L27)).

Would it be possible to expose the underlying providers in ChainedConfigurationProvider so that we can extract AzureAppConfigurationProvider and enable users to leverage our dynamic refresh functionality?

### PR DESCRIPTION
**Pull request by maryamariyan**
Fixes #44486

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration/src/ChainedConfigurationProvider.cs [hunk 1]
```diff
@@ -30,6 +30,11 @@ public ChainedConfigurationProvider(ChainedConfigurationSource source!!)
             _shouldDisposeConfig = source.ShouldDisposeConfiguration;
         }
 
+        /// <summary>
+        /// The chained configuration.
+        /// </summary>
+        public IConfiguration Configuration => _config;
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration/src/ChainedConfigurationProvider.cs:36]: nit: can we change the constructor implementation from 

```C#
        public ChainedConfigurationProvider(ChainedConfigurationSource source!!)
        {
            if (source.Configuration == null)
            {
                throw new ArgumentException(SR.Format(SR.InvalidNullArgument, "source.Configuration"), nameof(source));
            }
            _config = source.Configuration;
``` 

to 

```C#
        public ChainedConfigurationProvider(ChainedConfigurationSource source!!)
        {
            var configuration = source.Configuration;
            if (configuration  == null)
            {
                throw new ArgumentException(SR.Format(SR.InvalidNullArgument, "source.Configuration"), nameof(source));
            }
            _config = configuration ;
````

The reason `ChainedConfigurationSource .Configuration`  is settable and possible to set it to null I guess between the time of checking and the time assigning it to _config.

**Review by tarekgh — APPROVED** [review]

**Comment by ericstj** [comment]
cc @zhenlan

**Review by eerhardt — APPROVED** [review]
LGTM.

File: src/libraries/Microsoft.Extensions.Configuration/src/ChainedConfigurationProvider.cs [hunk 2]
```diff
@@ -21,15 +21,20 @@ public class ChainedConfigurationProvider : IConfigurationProvider, IDisposable
...
         {
-            if (source.Configuration == null)
+            var configuration = source.Configuration;
+            if (configuration  == null)
             {
                 throw new ArgumentException(SR.Format(SR.InvalidNullArgument, "source.Configuration"), nameof(source));
             }
-
-            _config = source.Configuration;
+            _config = configuration;
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Configuration/src/ChainedConfigurationProvider.cs:29]:
```C#
_config = source.Configuration ?? throw new ArgumentException(SR.Format(SR.InvalidNullArgument, "source.Configuration"), nameof(source));
```
?

**Pull request merged**